test/SgidClient.spec.ts

```diff
@@ -0,0 +1,332 @@
+import { readFileSync } from 'fs'
+
+import SgidClient from '../src'
+
+import {
+  MOCK_ACCESS_TOKEN,
+  MOCK_API_VERSION,
+  MOCK_AUTH_CODE,
+  MOCK_AUTH_ENDPOINT,
+  MOCK_CLIENT_ID,
+  MOCK_CLIENT_PRIVATE_KEY,
+  MOCK_CLIENT_SECRET,
+  MOCK_HOSTNAME,
+  MOCK_REDIRECT_URI,
+  MOCK_SUB,
+  MOCK_USERINFO_PLAINTEXT,
+} from './mocks/constants'
+import {
+  tokenHandlerNoSub,
+  tokenHandlerNoToken,
+  userInfoHandlerMalformedData,
+  userInfoHandlerMalformedKey,
+  userInfoHandlerNoData,
+  userInfoHandlerNoKey,
+} from './mocks/handlers'
+import { server } from './mocks/server'
+
+const DEFAULT_SCOPE = 'myinfo.nric_number openid'
+const DEFAULT_RESPONSE_TYPE = 'code'
+
+describe('SgidClient', () => {
+  let client: SgidClient
+
+  beforeEach(() => {
+    client = new SgidClient({
+      clientId: MOCK_CLIENT_ID,
+      clientSecret: MOCK_CLIENT_SECRET,
+      privateKey: MOCK_CLIENT_PRIVATE_KEY,
+      redirectUri: MOCK_REDIRECT_URI,
+      hostname: MOCK_HOSTNAME,
+      apiVersion: MOCK_API_VERSION,
+    })
+  })
+
+  describe('constructor', () => {
+    it('should initialise correctly when a PKCS1 private key is provided', () => {
+      expect(client).toBeDefined()
+    })
+
+    it('should initialise correctly when a PKCS8 private key is provided', () => {
+      const pkcs8Key = readFileSync(
+        `${__dirname}/mocks/mockPrivateKeyPkcs8.pem`,
+      ).toString()
+
+      const client = new SgidClient({
+        clientId: MOCK_CLIENT_ID,
+        clientSecret: MOCK_CLIENT_SECRET,
+        privateKey: pkcs8Key,
+        redirectUri: MOCK_REDIRECT_URI,
+        hostname: MOCK_HOSTNAME,
+        apiVersion: MOCK_API_VERSION,
+      })
+
+      expect(client).toBeDefined()
+    })
+  })
```

Comment:
Can consider add another test case for constructor error? i.e. negative test case for when PKCS1/PKCS8 key is not provided. This allows us to see if constructor throws proper error message especially if this is a library that will be used by public.

Reply:
hmm okay so the constructor doesn't actually throw when the private key isn't provided. currently if the private key is not provided or somehow malformed, the library will only throw when eventually you try to decrypt a payload by calling `userinfo`.

if we want the library to throw, we can tackle that in a separate PR and add a test for it then. currently I'm focusing on testing existing functionality.